x/il/keeper/abci.go

```diff
@@ -51,30 +51,52 @@ func (k Keeper) EndBlocker(ctx sdk.Context) {
 			return false
 		}
 
-		reserveStr0 := strconv.FormatFloat(pair.Reserve0, 'f', 2, 64)
-		if reserveStr0 == "" {
+		// reserveStr0 := strconv.FormatFloat(pair.Reserve0, 'f', 2, 64)
+		// if reserveStr0 == "" {
+		// 	return false
+		// }
+
+		// reserveStr1 := strconv.FormatFloat(pair.Reserve1, 'f', 2, 64)
+		// if reserveStr1 == "" {
+		// 	return false
+		// }
+
+		totalSupplyStr := strconv.FormatFloat(pair.TotalSupply, 'f', 2, 64)
+		if totalSupplyStr == "" {
 			return false
 		}
 
-		reserveStr1 := strconv.FormatFloat(pair.Reserve1, 'f', 2, 64)
-		if reserveStr1 == "" {
+		reserveUSDStr := strconv.FormatFloat(pair.ReserveUsd, 'f', 2, 64)
+		if reserveUSDStr == "" {
 			return false
 		}
 
-		reserve0, err := sdk.NewDecFromStr(reserveStr0)
+		// reserve0, err := sdk.NewDecFromStr(reserveStr0)
+		// if err != nil {
+		// 	return false
+		// }
+
+		// reserve1, err := sdk.NewDecFromStr(reserveStr1)
+		// if err != nil {
+		// 	return false
+		// }
+
+		totalSupply, err := sdk.NewDecFromStr(totalSupplyStr)
 		if err != nil {
 			return false
 		}
 
-		reserve1, err := sdk.NewDecFromStr(reserveStr1)
+		reserveUSD, err := sdk.NewDecFromStr(reserveUSDStr)
```

Comment:
Sorry to go back and forth on this 😀 but is there a reason we need to support one or the other (one being 'price ratio', the other being 'usd value'). Can we support both?

Don't know how the rest of the protocol works but I'm assuming this would require a new field on `stoploss`, something like `stoploss.calculationType`. That could have values of `priceRatio` or `usdValue`. Then depending on the value of `calculationType` you would calculate slippage based on price ratio or based on usd value.

My follow-up question would be - let's say your stoploss is at 10%. Does a 10% change in the price ratio of the two pairs equate to a 10% loss of usd value of the position? Definitely not, especially if neither token in the pair is a stablecoin. 

Some back of the napkin math...
```
1 WETH to 1000 USDC - 2000 USD liquidity
…WETH price goes up to 1500 USDC….
Price ratio is 1500 WETH to 1 USDC - 50% price ratio change
This is a 2% IL loss (realized loss is smaller according to fees)
```

And we can see from https://uniswap.org/docs/v2/advanced-topics/understanding-returns/ that the relationship between price ratio and IL is plotted by this graph:

<img width="740" alt="Screen Shot 2021-02-08 at 11 06 45 AM" src="https://user-images.githubusercontent.com/12480403/107269197-17647000-69fe-11eb-8a34-4c37469f812e.png">

So, question is - can we give the user a way to model their loss in _either_ way - price ratio OR usd value - and then use the relationship in the screenshot above to figure out their loss in their preferred model?

Don't want to overcomplicate things, if we have to pick one approach, I think USD value is best, so this code looks good. But wondering if this type of flexibility is possible b/c that's how I was thinking about the UI interaction, the user would get a chance to say either:

* 'stop me out when I lose X USD' OR
* 'stop me out when I get rebalanced Y far into one coin' (this is equivalent to price ratio)

And the modeled relationship above is what I think we need to achieve that

Reply:
Basically this is the relationship:

```
initialPriceRatio = stoploss.ReferencePairRatio
currentPriceRatio = pair.reserve0 / pair.reserve1
priceRatioChange = initialPriceRatio / currentPriceRatio
impermanentLossRatio = 2 * sqrt(priceRatioChange) / (1+priceRatioChange) - 1 
impermanentLoss = impermanentLossRatio * usdValueOfPosition
totalReturn = stoploss.accruedFees - impermanentLoss (not sure if we have accrued fees)
```

Then, depending on how the user chose to model it, we either see if totalReturn is less than 90% of their initial usd value of their position, or if priceRatioChange is over 10%. (assuming 10% as the 'stoploss baseline').

Reply:
Thanks @kkennis for the detailed comment

> -  'stop me out when I lose X USD' OR
- 'stop me out when I get rebalanced Y far into one coin' (this is equivalent to price ratio)

I think the first one is what users would care about the most in the beginning. But I'm def open to integrate both options, but I'd prefer to do the second one once wire the IL module with the smart contract and oracle logic and refactor the oracle module. Would that be ok? I'm open to suggestions